src/fairseq2/utils/rng.py

```diff
@@ -40,3 +26,90 @@ def use_deterministic(value: bool, warn_only: bool = False) -> None:
     torch.backends.cudnn.benchmark = not value
 
     torch.use_deterministic_algorithms(value, warn_only=warn_only)
+
+
+class GeneratorBag:
+    """Holds a collection of random number generators."""
+
+    generators: List[Generator]
+
+    def __init__(self, *generators: Generator) -> None:
+        """
+        :param generators:
+            The generators to hold.
+        """
+        self.generators = list(generators)
+
+    @staticmethod
+    def from_device_defaults(*devices: Device) -> GeneratorBag:
+        """Create a :class:`GeneratorBag` instance holding the default random
+        number generators of ``devices``."""
+        generators = []
+
+        for device in set(devices):
+            if device.type == "cpu":
+                generators.append(torch.default_generator)
+            elif device.type == "cuda":
+                # Ensure that the default CUDA generators are initialized.
+                torch.cuda.init()
```

Comment:
Curious why do we need this - doesnt this happen automatically ? Or is it related to the use of fairseqn

Reply:
The initialization happens automatically once you call a CUDA op, but, for instance, does not happen if your Python script starts with:

```
d = torch.device("cuda:0")

bag = RNGBag.from_default_devices(d)  # raises IndexError
```

Since we haven't called any CUDA op till this point, `torch.cuda.default_generators` will be empty. In vast majority of the use cases in the real world, CUDA will be already initialized by the time a function such as this will be called, but there is no guarantee. So I included this call to ensure it.

Reply:
Thank you for the explanation!